test: Signature Redesign e2e tests for each supported signature type

digiwand opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24352?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/23977
Blocked By: https://github.com/MetaMask/metamask-extension/issues/24712

## **Manual testing steps**


1. set `ENABLE_CONFIRMATION_REDESIGN=true` in .metamaskrc
2. enable Experimental > "Improved signature requests" in settings
3. test signatures in [test-dapp](https://metamask.github.io/test-dapp/)

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

digiwand committed:
refactor: ConfirmFooter data-testid "confirm-footer-button"

digiwand committed:
test:e2e: signature redesign personal_sign

digiwand committed:
test:e2e: signature eth_signTypedData

digiwand committed:
test:refactor: '127.0.0.1:8080' -> DAPP_URL_WITHOUT_SCHEMA

digiwand committed:
test:refactor require -> import

digiwand committed:
test:fix: require -> import

digiwand committed:
test:refactor: snake case -> kebab case filenames

digiwand committed:
test:fix: require FixtureBuilder

digiwand committed:
test:e2e signature redesign deprecate isElementPresent

digiwand committed:
test: signature redesign reorganize imports

digiwand committed:
test:fix: flaky personalSignVerifyECRecoverResult

digiwand committed:
test:e2e add signature redesign signTypedDataV3 test

digiwand committed:
test:e2e: signature redesign update test phrase

digiwand committed:
refactor: redesign signTypedDataV3 isElementPresent -> findElement

digiwand committed:
test:e2e: create signature redesign signTypedDataV4

digiwand committed:
test:e2e: create signature redesign permit

digiwand committed:
refactor: clean whitespaces

digiwand committed:
test: create and use withRedesignConfirmationFixtures

digiwand committed:
test: fix whitespace & newlines manually
- looks like eslint it not setup for these files

digiwand committed:
test:e2e: add signature redesign reject tests

digiwand committed:
fix:test: Signature Redesign footer and page snapshots which include data-testId

seaona commented on test/e2e/tests/confirmations/helper-fixture.ts:
```diff
@@ -0,0 +1,25 @@
+    {
+      dapp: true,
+      fixtures: new FixtureBuilder()
+      .withNetworkControllerOnMainnet()
```
hi there! nice file addition :fire: 
just wondering why do we need to use Mainnet here? I think if it's not strictly needed, we could use the local network by default (ganache) in all these specs

seaona commented on test/e2e/tests/confirmations/signatures/personal-sign.spec.ts:
```diff
@@ -0,0 +1,76 @@
+import { Suite } from 'mocha';
+import { withRedesignConfirmationFixtures } from '../helper-fixture';
+import {
+  DAPP_URL_WITHOUT_SCHEMA,
```
nice

digiwand committed:
fix:test: rn DAPP_URL_WITHOUT_SCHEMA -> DAPP_HOST_ADDRESS

digiwand committed:
fix:test: update signature design after removing mainnet network fixture

digiwand committed:
test:e2e:fix: flaky Signature Request rejection tests

pedronfigueiredo commented on test/e2e/tests/confirmations/signatures/permit.spec.ts:
```diff
@@ -0,0 +1,112 @@
+        await driver.clickElement('#signPermit');
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+
+        const origin = driver.findElement({ text: DAPP_HOST_ADDRESS });
+        const contractPetName = driver.findElement({
+          css: '.name__value',
+          text: '0xCcCCc...ccccC',
+        });
+
+        const primaryType = driver.findElement({ text: 'Permit' });
+        const owner = driver.findElement({ css: '.name__name', text: 'Account 1' });
+        const spender = driver.findElement({ css: '.name__value', text: '0x5B38D...eddC4' });
+        const value = driver.findElement({ text: '3000' });
+        const nonce = driver.findElement({ text: '0' });
+        const deadline = driver.findElement({ text: '50000000000' });
+
+        assert.ok(await origin, 'origin');
+        assert.ok(await contractPetName, 'contractPetName');
+
+        assert.ok(await primaryType, 'primaryType');
+        assert.ok(await owner, 'owner');
+        assert.ok(await spender, 'spender');
+        assert.ok(await value, 'value');
+        assert.ok(await nonce, 'nonce');
+        assert.ok(await deadline, 'deadline');
```
Perhaps we can extract this to another assert method?

digiwand replied:
Possibly. Is this to clean the function and have shorter function lengths, or a different specific asset method? cc: @pedronfigueiredo

pedronfigueiredo replied:
I'm not sure what you mean by asset method in this context, but this is what I had in mind

```js
async function assertSignatureElementsRendered(driver) {
        const origin = driver.findElement({ text: DAPP_HOST_ADDRESS });
        assert.ok(await origin, 'origin');

        const contractPetName = driver.findElement({
            css: '.name__value',
            text: '0xCcCCc...ccccC',
        });
        assert.ok(await contractPetName, 'contractPetName');

        const primaryType = driver.findElement({ text: 'Permit' });
        assert.ok(await primaryType, 'primaryType');

        const owner = driver.findElement({ css: '.name__name', text: 'Account 1' });
        assert.ok(await owner, 'owner');

        const spender = driver.findElement({ css: '.name__value', text: '0x5B38D...eddC4' });
        assert.ok(await spender, 'spender');

        const value = driver.findElement({ text: '3000' });
        assert.ok(await value, 'value');

        const nonce = driver.findElement({ text: '0' });
        assert.ok(await nonce, 'nonce');

        const deadline = driver.findElement({ text: '50000000000' });
        assert.ok(await deadline, 'deadline');
}
```

digiwand replied:
I see. Okay can do 👍🏼

pedronfigueiredo replied:
Interesting, I wasn't aware of that, but I wonder if there's enough of a difference to justify prioritizing that over optimal code readability/organization

pedronfigueiredo replied:
This is a nit, feel free to ignore that part of the suggestion if it makes more sense to you

digiwand replied:
actually, you are right. performance is negligible in this case. time 24 → time 25. I believe I realized this while testing elements that rendered with a delay in the test-dapp.  

In regards to readability, I don't quite agree yet the way you wrote it is more readable. I also wrote the code with the same concern for readability in mind. I found it easier skim the function when the method names aligned. The minimal delta in line lengths in consecutive order aids as well. With this, we also can understand what we are asserting by reading half of the lines rather than all lines of the function. I agree it comes with a cost of separating the const with the use in assert. Happy to change this if you or others still feel strongly about the update
<img width="747" alt="Screenshot 2024-05-07 at 7 06 46 PM" src="https://github.com/MetaMask/metamask-extension/assets/20778143/6e061d20-5e46-47c1-9e41-b1c1ff406d9f">

pedronfigueiredo commented on test/e2e/tests/confirmations/signatures/personal-sign.spec.ts:
```diff
@@ -0,0 +1,79 @@
+        await driver.clickElement('#personalSign');
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+
+        const origin = driver.findElement({ text: DAPP_HOST_ADDRESS });
+        const message = driver.findElement({ text: 'Example `personal_sign` message' });
+
+        assert.ok(await origin);
+        assert.ok(await message);
```
same as https://github.com/MetaMask/metamask-extension/pull/24352/files#r1589372927

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand committed:
fix: rn preference redesignedConfirmationsEnabled -> redesignedConfirmations following latest merge

digiwand committed:
test:e2e:refactor: extract info value assets into separate function

digiwand committed:
test:fix: confirm redesign personal-sign reject test

digiwand committed:
test:e2e: potentially fix flaky test

digiwand committed:
test:e2e: potentially fix flaky tests 2

digiwand committed:
test:e2e:fix: rm unused const

matthewwalsh0 commented on test/e2e/helpers.js:
```diff
@@ -643,6 +643,7 @@ const closeSRPReveal = async (driver) => {
 };
 
+const DAPP_HOST_ADDRESS = '127.0.0.1:8080';
 const DAPP_URL = 'http://127.0.0.1:8080';
```
Should we update this constant to use the above to keep them aligned?

digiwand replied:
I do prefer to leave it as is since it's concise, rather than updating it to e.g. `DAPP_HOST_ADDRESS_WITH_SCHEMA`, as prepending the host address with the schema turns the value into a URL

matthewwalsh0 replied:
Sorry for the confusion, I meant should we update `DAPP_URL` to use `DAPP_HOST_ADDRESS`.

digiwand replied:
ooh, sure! can do. updated b05d8569b5884ba20d196bc1cea5c74b79149341

matthewwalsh0 commented on test/e2e/tests/confirmations/signatures/personal-sign.spec.ts:
```diff
@@ -0,0 +1,80 @@
+
+        await assertInfoValues(driver);
+
+        await driver.clickElement('[data-testid="confirm-footer-button"]');
```
Minor, but is it worth creating a helper file in this directory to provide reusable functions such as `clickConfirmButton` as we do this in many of the tests?

It also helps to keep the test functions themselves high level and readable, plus it moves us slightly closer to page model objects that I'm a big fan of.

digiwand replied:
hm.. unsure what helpers you are refering to. If it is what I am thinking - I am more of a fan of creating consts to reference the selectors unless the helper functions would involve more than one line

cryptotavares replied:
I think that @matthewwalsh0 is hinting at some kind of page object model. We do not have any specific guidelines in the Extension... but there are some helper functions that are added [here](https://github.com/MetaMask/metamask-extension/blob/develop/test/e2e/helpers.js).

matthewwalsh0 replied:
As discussed, not urgent so can be done in future.

matthewwalsh0 commented on test/e2e/tests/confirmations/signatures/sign-typed-data-v4.spec.ts:
```diff
@@ -0,0 +1,107 @@
+
+        await unlockWallet(driver);
+        await openDapp(driver);
+        await driver.clickElement('#signTypedDataV4');
```
Is everything up to this line the same in every test in every file?

Should we create a `createSignatureConfirmation` function in that helper file I mentioned in the other comment?

It could accept an enum such as `SignatureType` to encapsulate the button ID we need to click also.

digiwand replied:
I like this idea. Could be used outside of the confirmation redesign files as well. Created an issue ticket to handle this after the release https://github.com/MetaMask/metamask-extension/issues/24468

matthewwalsh0 commented on test/e2e/tests/confirmations/signatures/sign-typed-data-v4.spec.ts:
```diff
@@ -0,0 +1,107 @@
+import { Driver } from '../../../webdriver/driver';
+
+describe('Confirmation Signature - Sign Typed Data V4', function (this: Suite) {
+  if (!process.env.ENABLE_CONFIRMATION_REDESIGN) { return; }
```
Minor, but even this env check could be behind a helper function such as `isRedesignEnabled` so we don't have to spread the knowledge of the ENV.

digiwand replied:
hm.. interesting. this would be a new pattern. I'm content with checking the ENV directly. If we were to do this, and with consideration of the `redesignedConfirmationsEnabled` selector naming, we could add the check to `shared/modules/environment.ts` as

`const isEnvEnableConfirmationRedesign = process.env.ENABLE_CONFIRMATION_REDESIGN`

would this proposed update be preferred? @matthewwalsh0

digiwand replied:
actually, with the following PR that'll be merged soon, we'll remove the ENV checks in this PR 
https://github.com/MetaMask/metamask-extension/pull/24466

cryptotavares replied:
Yes we will remove the env (probably before this PR gets merged 😅)

matthewwalsh0 commented on test/e2e/tests/confirmations/helper-fixture.ts:
Would this file be a good candidate for the helper functions, we could just call it `helper.ts` to be more general?

matthewwalsh0 commented on test/e2e/tests/confirmations/signatures/sign-typed-data-v4.spec.ts:
```diff
@@ -0,0 +1,107 @@
+  const fromName = driver.findElement({ text: 'Cow' });
+  const fromAddressNum0 = driver.findElement({ css: '.name__value', text: '0xCD2a3...DD826' });
+  const toName = driver.findElement({ text: 'Bob' });
+  const toAddressNum2 = driver.findElement({ css: '.name__value', text: '0xB0B0b...00000' });
```
Could this be abstracted to a reusable `findName` across the files?

digiwand replied:
what if we create a resuable const?
```
const selectorPetName = '.name__value'
```

This way, when reading the the line, we understand what it is doing. Not sure if it's worth needing to view `findName` or `findElementByPetName` to understand the logic

matthewwalsh0 replied:
This is part of the abstractions so not urgent and can be included in the future work.

matthewwalsh0 commented on test/e2e/tests/confirmations/signatures/sign-typed-data-v4.spec.ts:
```diff
@@ -0,0 +1,107 @@
+    text: '0xCcCCc...ccccC',
+  });
+
+  const primaryType = driver.findElement({ text: 'Mail' });
```
We're currently included the constants directly in the assertions, but is it worth the decoupling any perhaps readability to create a more generic `expectV4Signature` that accepts all the values as properties and we can instead define the values as constants?

We could mirror this in the other files also.

matthewwalsh0 replied:
Exactly, just an `expectV4Signature` method that accepts an object of the expected properties to verify.

But again future abstractions that aren't crucial now given timelines.

jpuri commented:
Possible to loom through various signatures, rather than adding a separate spec for each ?

digiwand commented:
Hi @jpuri, this was a consideration. Discussed this with @pedronfigueiredo and @pnarayanaswamy. We've been leaning towards keeping files per type. I'm open to other considerations. This said, @pnarayanaswamy has created a followup PR to refactor our tests and combine them together here https://github.com/MetaMask/metamask-extension/pull/24424

digiwand committed:
test:e2e:refactor: rn confirm-redesign helper-fixture.ts -> herlper.ts

digiwand committed:
test:refactor: use DAPP_HOST_ADDRESS in DAPP_URL

cryptotavares reviewed (dismissed):
We probably can improve the abstractions as @matthewwalsh0 is suggesting, but we should get define some guidelines on how to do so (nudging you @seaona 😅). I don't think that that's in the scope of this PR. So lgtm 🟢

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand commented on test/e2e/tests/confirmations/signatures/sign-typed-data-v4.spec.ts:
```diff
@@ -0,0 +1,107 @@
+
+  assert.equal(await verifyResult.getText(), '0xcd2f9c55840f5e1bcf61812e93c1932485b524ca673b36355482a4fbdf52f692684f92b4f4ab6f6c8572dacce46bd107da154be1c06939b855ecce57a1616ba71b');
+  assert.equal(await verifyRecoverAddress.getText(), publicAddress);
+}
```
the confirmation-redesign test builds seem to be flaky 🤔  
![CleanShot 2024-05-09 at 21 55 31@2x](https://github.com/MetaMask/metamask-extension/assets/20778143/697bb83e-eb46-40c4-896c-5e2e2e6a1085)

pnarayanaswamy replied:
this is because sometimes there is a delay in the dialog opening up. switch  `await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);` to the method  `await switchToNotificationWindow(driver);` This should make it less flaky 🤞

digiwand replied:
ohh I see now. Thank you @pnarayanaswamy! updated [bd82ec6](https://github.com/MetaMask/metamask-extension/pull/24352/commits/bd82ec636ddb44f24eefc41f0dd145ba8a887526)

digiwand committed:
test:e2e:fix: confirm redesign TypeError: windowHandles is not iterable

digiwand committed:
test:fix: flaky signature redesign w/ waitUntilXWindowHandles

digiwand committed:
test:e2e: fix flaky signature redesign w/ switchToNotificationWindow

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand committed:
test:fix: flaky test pt.3 use both waitUntilXWindowHandles & switchToNotificationWindow

digiwand committed:
test:fix: attempt to fix flaky test with driverOptions: timeOut

digiwand committed:
test:e2e:refactor: rn
 confirmations helper -> helpers

digiwand committed:
test:e2e:fix: confirm redesign scrollable confirmations

digiwand committed:
test:e2e:refactor: confirm redesign rm unused ganacheServer in reject tests

digiwand commented:
somehow, this unit test failed in this PR
https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/80783/workflows/9be32d3b-80da-41b8-8cfd-1d837f57871b/jobs/2855569

![CleanShot 2024-05-15 at 15 56 28@2x](https://github.com/MetaMask/metamask-extension/assets/20778143/7560d512-189a-424f-a208-809439179d51)

we should double-check this is not a flaky test separately from this PR cc: @jpuri

digiwand commented:
https://app.circleci.com/jobs/github/MetaMask/metamask-extension/2855977

failing on the chrome-confirmation-redesign builds 🤔 
```
Warning: vkCreateInstance: Found no drivers!
Warning: vkCreateInstance failed with VK_ERROR_INCOMPATIBLE_DRIVER
    at CheckVkSuccessImpl (../../third_party/dawn/src/dawn/native/vulkan/VulkanError.cpp:101)
```

![CleanShot 2024-05-15 at 16 13 07@2x](https://github.com/MetaMask/metamask-extension/assets/20778143/acd7c6da-8628-41e0-9a40-5752a775e6bd)
![CleanShot 2024-05-15 at 16 59 00@2x](https://github.com/MetaMask/metamask-extension/assets/20778143/114ac635-4063-417c-a74c-a777118d12b0)

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand committed:
feat: finish merge conflict

digiwand committed:
fix: CircleCI test-e2e-chrome-confirmation-redesign job (#24684)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

created script to build script with ENABLE_CONFIRMATION_REDESIGN flag
and added to test-e2e-chrome-confirmation-redesign command

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24684?quickstart=1)

## **Related issues**
Fixes: https://github.com/MetaMask/metamask-extension/issues/24712
Fixes: https://github.com/MetaMask/metamask-extension/issues/23975
Continues and related to:
https://github.com/MetaMask/metamask-extension/pull/24240

## **Manual testing steps**

check ci/circleci: test-e2e-chrome-confirmation-redesign job

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Priya Narayanaswamy <priya.narayanaswamy@consensys.net>

digiwand committed:
test:e2e: rm getAttribute asserts for Confirmations scroll helper
- asynchronous thenable that is not awaited is not being tested

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand commented:
resolved a 1 line merge conflict f53b310acf5dc5fe79cfcdb3c2348f9e5b0e8d16. please rereview @pedronfigueiredo @matthewwalsh0 🙏🏼

jpuri commented on test/e2e/webdriver/driver.js:
```diff
@@ -159,6 +159,7 @@ class Driver {
   }
 
   buildLocator(locator) {
+    console.log('buildLocator: ', locator);
```
🔪

digiwand replied:
```suggestion
```

digiwand replied:
thanks for the catch! Removed

digiwand committed:
chore: rm console.log

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

digiwand commented:
resolved another minor merge conflict [fe81382](https://github.com/MetaMask/metamask-extension/pull/24352/commits/fe81382a81c57478b8e818fbeb2e5ce7a31add76)

digiwand committed:
Merge branch 'develop' into test-e2e-signature-redesign

danjm reviewed (changes requested):
Sorry to block on such a small thing :sweat_smile: 

but this was intentionally changed to 1 very recently https://github.com/MetaMask/metamask-extension/pull/24531

digiwand commented on .circleci/config.yml:
```diff
@@ -1189,7 +1189,7 @@ jobs:
           command: |
             if .circleci/scripts/test-run-e2e.sh
             then
-              timeout 20m yarn test:e2e:chrome --retries 1
```
good catch. Ty!

digiwand replied:
updated [6e183cb](https://github.com/MetaMask/metamask-extension/pull/24352/commits/6e183cbe7ccbb65753f02ab9908914bfae313f02)

digiwand commented on .circleci/config.yml:
```diff
@@ -1189,7 +1189,7 @@ jobs:
             if .circleci/scripts/test-run-e2e.sh
             then
-              timeout 20m yarn test:e2e:chrome --retries 1
+              timeout 20m yarn test:e2e:chrome:confirmation-redesign --retries 2
```
```suggestion
              timeout 20m yarn test:e2e:chrome:confirmation-redesign --retries 1
```

digiwand committed:
fix: set confirmation e2e circleci retry count to 1

danjm reviewed (approved):
Approving .circleci/config.yml changes on behalf of library admins group
